include/sys/zio_compress.h

```diff
@@ -51,15 +53,83 @@ enum zio_compress {
 	ZIO_COMPRESS_GZIP_9,
 	ZIO_COMPRESS_ZLE,
 	ZIO_COMPRESS_LZ4,
+	ZIO_COMPRESS_ZSTD,
 	ZIO_COMPRESS_FUNCTIONS
 };
 
+#define	ZIO_COMPLEVEL_INHERIT	0
+#define	ZIO_COMPLEVEL_DEFAULT	255
+
+enum zio_zstd_levels {
+	ZIO_ZSTDLVL_INHERIT = 0,
+	ZIO_ZSTDLVL_1,
+#define	ZIO_ZSTD_LEVEL_MIN	ZIO_ZSTDLVL_1
+	ZIO_ZSTDLVL_2,
+	ZIO_ZSTDLVL_3,
+#define	ZIO_ZSTD_LEVEL_DEFAULT	ZIO_ZSTDLVL_3
+	ZIO_ZSTDLVL_4,
+	ZIO_ZSTDLVL_5,
+	ZIO_ZSTDLVL_6,
+	ZIO_ZSTDLVL_7,
+	ZIO_ZSTDLVL_8,
+	ZIO_ZSTDLVL_9,
+	ZIO_ZSTDLVL_10,
+	ZIO_ZSTDLVL_11,
+	ZIO_ZSTDLVL_12,
+	ZIO_ZSTDLVL_13,
+	ZIO_ZSTDLVL_14,
+	ZIO_ZSTDLVL_15,
+	ZIO_ZSTDLVL_16,
+	ZIO_ZSTDLVL_17,
+	ZIO_ZSTDLVL_18,
+	ZIO_ZSTDLVL_19,
+#define	ZIO_ZSTD_LEVEL_MAX	ZIO_ZSTDLVL_19
+#define	ZIO_ZSTDLVL_MAX		ZIO_ZSTDLVL_19
+	ZIO_ZSTDLVL_RESERVE = 101, /* Leave room for new positive levels */
+	ZIO_ZSTDLVL_FAST, /* Fast levels are negative */
+	ZIO_ZSTDLVL_FAST_1,
+#define	ZIO_ZSTD_FAST_LEVEL_DEFAULT	ZIO_ZSTDLVL_FAST_1
+	ZIO_ZSTDLVL_FAST_2,
+	ZIO_ZSTDLVL_FAST_3,
+	ZIO_ZSTDLVL_FAST_4,
+	ZIO_ZSTDLVL_FAST_5,
+	ZIO_ZSTDLVL_FAST_6,
+	ZIO_ZSTDLVL_FAST_7,
+	ZIO_ZSTDLVL_FAST_8,
+	ZIO_ZSTDLVL_FAST_9,
+	ZIO_ZSTDLVL_FAST_10,
+	ZIO_ZSTDLVL_FAST_20,
+	ZIO_ZSTDLVL_FAST_30,
+	ZIO_ZSTDLVL_FAST_40,
+	ZIO_ZSTDLVL_FAST_50,
+	ZIO_ZSTDLVL_FAST_60,
+	ZIO_ZSTDLVL_FAST_70,
+	ZIO_ZSTDLVL_FAST_80,
+	ZIO_ZSTDLVL_FAST_90,
+	ZIO_ZSTDLVL_FAST_100,
+	ZIO_ZSTDLVL_FAST_500,
+	ZIO_ZSTDLVL_FAST_1000,
+#define	ZIO_ZSTDLVL_FAST_MAX	ZIO_ZSTDLVL_FAST_1000
+	ZIO_ZSTDLVL_DEFAULT = 250, /* Allow the default level to change */
```

Comment:
Is it not confusing to define both `ZIO_ZSTDLVL_DEFAULT` here (which I guess is the "internal" compression level default as the ZSTD algo understands things) as well as `ZIO_ZSTD_LEVEL_DEFAULT` above on line 69 (which is the default for the "semantic" or "user-facing" level option ZFS is presenting to users from 1-19)? I know the former is an enum and the latter is a constant, but my mind parses these the same without doing a double-take and I think that's a recipe for bugs, especially since their values are both integers.
Same criticism applies to lines 86 and 87, `ZIO_ZSTD_LEVEL_MAX` and `ZIO_ZSTDLVL_MAX`, in those cases they're both consts and seem to mean the same thing semantically, yet they're BOTH defined? This seems silly. I'd stick to 1 naming convention for the enums, and a distinctly different naming convention for the consts, throughout.

Martin Fowler was right about the 2 hard things in computer programming, and naming correctly is the underrated one IMHO!

(sorry for the drive-by, just a relatively new ZFS user and dev who is curious about this feature!)

Reply:
Oh, indeed that should be changed!

> sorry for the drive-by

No apology necessary! Thank you very much :)

Reply:
There isn't even a reason the make a difference between the defines and enums, so let's just keep one scheme

Reply:
There's another possible destructive-interaction consequence of relying on auto-value-assigned enums when it comes to feature flags combined with inserting new enums later on (which shuffles the auto-assigned values of later enums up by 1), I'm not certain if ZFS does this precisely but I did see that it uses feature flags, you can probably see the consequence of generating a bitmask via OR'ing a bunch of features like (`1 << ZIO_insert_feature_here`), and then assigning that feature-flag int to a pool, and then later adding a new enum value to the ZFS code above that feature (like has been done in this PR), now the feature bit on the feature value in the pool will be read wrong by the new ZFS version because it's now slotted to the wrong bit... 🙅‍♂ 
The upshot is that I think manually assigning an int value to every enum is more deterministic and less error-prone, even if it's more of a pain upfront, because it means that adding enums later won't have possible unforeseen consequences in code which might assume the actual value of the enum is immutable. The only catch is that your initial explicit value assignment ideally MUST correspond to how the compiler is already assigning the values (in order to not create a new bug just by making things explicit!)
Happy to help, I like code reviews for some reason ¯\_(ツ)_/¯

Reply:
From @BrainSlayer:

> @pmarreck i agree with your proposal to change the naming convention. even if there is no much space for doing misstakes here since its not used wide spread over the code, but just on 2 single locations (the mapping between these values)